// @cognee/cognee-ts/lib/cancellation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CancellationToken = exports.CancellationHandle = void 0;
exports.createCancellationPair = createCancellationPair;
const native_1 = require("./native");
class CancellationHandle {
    /** @internal */
    constructor(box_) {
        this._box = box_;
    }
    cancel() {
        native_1.native.cancellationHandleCancel(this._box);
    }
    get isCancelled() {
        return native_1.native.cancellationHandleIsCancelled(this._box);
    }
    clone() {
        return new CancellationHandle(native_1.native.cancellationHandleClone(this._box));
    }
}
exports.CancellationHandle = CancellationHandle;
class CancellationToken {
    /** @internal */
    constructor(box_) {
        this._box = box_;
    }
    get isCancelled() {
        return native_1.native.cancellationTokenIsCancelled(this._box);
    }
    clone() {
        return new CancellationToken(native_1.native.cancellationTokenClone(this._box));
    }
}
exports.CancellationToken = CancellationToken;
/** Create a linked (handle, token) cancellation pair. */
function createCancellationPair() {
    const result = native_1.native.cancellationPair();
    return {
        handle: new CancellationHandle(result.handle),
        token: new CancellationToken(result.token),
    };
}
