// @cognee/cognee-ts/lib/cognee.js
"use strict";
/**
 * High-level TypeScript SDK for the cognee AI-memory pipeline.
 *
 * ```ts
 * import { Cognee } from 'cognee';
 *
 * const c = new Cognee({ llmModel: "gpt-4o-mini" });
 * await c.warm();
 * await c.add({ type: "text", text: "Hello, world!" }, "main");
 * await c.cognify("main");
 * const results = await c.search("Hello");
 * ```
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Cognee = void 0;
const native_1 = require("./native");
const errors_1 = require("./errors");
// Cloud module-level functions (`serve` / `disconnect`) live in the closed
// `cognee-ts-cloud` package (T15e). The OSS `cognee` package does not
// expose them.
/** Convert a single `snake_case` key to `camelCase`. */
function snakeToCamel(key) {
    return key.replace(/_([a-z0-9])/g, (_m, c) => c.toUpperCase());
}
/**
 * Rewrite the top-level keys of a config snapshot from `snake_case` (the wire
 * shape returned by the native `getConfig`) to the documented `camelCase` API
 * keys (`llm_model` -> `llmModel`, `chunk_size` -> `chunkSize`, ...), so the
 * object returned by `config.get()` matches `CogneeConfigObject` / the setters.
 */
function configToCamel(obj) {
    const out = {};
    for (const [k, v] of Object.entries(obj)) {
        out[snakeToCamel(k)] = v;
    }
    // The `setEmbeddingModel` setter writes the Rust `embedding_model_name`
    // field, so the snapshot surfaces it as `embeddingModelName`. Expose an
    // `embeddingModel` alias too so the read-back key matches the setter name.
    if ("embeddingModelName" in out && !("embeddingModel" in out)) {
        out.embeddingModel = out.embeddingModelName;
    }
    return out;
}
/** Convert a single `camelCase` key to `snake_case`. */
function camelToSnake(key) {
    return key.replace(/[A-Z]/g, (c) => `_${c.toLowerCase()}`);
}
/**
 * Public camelCase setting names whose `snake_case` form does NOT match the
 * native `Settings` field. Keyed by the camelCase name the SDK documents
 * (matching the `config.set*` setters and `config.get()` read-back); the value
 * is the actual `Settings` field. Only `embeddingModel` differs (the field is
 * `embedding_model_name`, mirroring `setEmbeddingModel`).
 */
const SETTINGS_KEY_ALIASES = {
    embeddingModel: "embedding_model_name",
};
/**
 * Normalize a constructor settings object so the documented camelCase keys
 * (`llmModel`, `embeddingProvider`, `chunkSize`, ...) reach the native
 * snake_case `Settings` deserializer. snake_case keys pass through unchanged
 * (camel→snake is idempotent on them), so both spellings are accepted — fixing
 * silently-ignored camelCase constructor args.
 */
function normalizeSettings(settings) {
    const out = {};
    for (const [k, v] of Object.entries(settings)) {
        const key = SETTINGS_KEY_ALIASES[k] ?? camelToSnake(k);
        out[key] = v;
    }
    return out;
}
/**
 * The main Cognee SDK class.
 *
 * Construct with an optional settings object (or JSON string) to override
 * env-derived defaults. Settings keys are the canonical `Settings` / env-var
 * field names (`llmModel`, `embeddingProvider`, etc.).
 *
 * ```ts
 * import { Cognee } from 'cognee';
 *
 * const c = new Cognee({ llmModel: "gpt-4o-mini", llmApiKey: process.env.OPENAI_API_KEY });
 * await c.warm();
 * await c.add({ type: "text", text: "The quick brown fox" }, "demo");
 * await c.cognify("demo");
 * const answer = await c.search("What does the fox do?");
 * ```
 */
class Cognee {
    constructor(settings) {
        // Accept the documented camelCase setting keys (e.g. `{ llmModel: ... }`)
        // as well as raw snake_case. The native `cogneeNew` deserializes a
        // snake_case `Settings`, so camelCase keys would otherwise be silently
        // ignored. A JSON string is parsed, normalized, and forwarded as an object.
        let normalized = settings;
        if (settings != null) {
            const obj = typeof settings === "string" ? JSON.parse(settings) : settings;
            normalized = normalizeSettings(obj);
        }
        this._handle = native_1.native.cogneeNew(normalized);
        // ── config sub-object ────────────────────────────────────────────────────
        // Granular setters are sync and do not need try/catch.
        // Fallible setters (configSet, bulk setters) are wrapped.
        const h = this._handle;
        this.config = {
            // LLM
            setLlmProvider: (v) => { native_1.native.configSetLlmProvider(h, v); },
            setLlmModel: (v) => { native_1.native.configSetLlmModel(h, v); },
            setLlmApiKey: (v) => { native_1.native.configSetLlmApiKey(h, v); },
            setLlmEndpoint: (v) => { native_1.native.configSetLlmEndpoint(h, v); },
            setLlmApiVersion: (v) => { native_1.native.configSetLlmApiVersion(h, v); },
            setLlmTemperature: (v) => { native_1.native.configSetLlmTemperature(h, v); },
            setLlmStreaming: (v) => { native_1.native.configSetLlmStreaming(h, v); },
            setLlmMaxCompletionTokens: (v) => { native_1.native.configSetLlmMaxCompletionTokens(h, v); },
            setLlmMaxRetries: (v) => { native_1.native.configSetLlmMaxRetries(h, v); },
            setLlmMaxParallelRequests: (v) => { native_1.native.configSetLlmMaxParallelRequests(h, v); },
            // Embedding
            setEmbeddingProvider: (v) => { native_1.native.configSetEmbeddingProvider(h, v); },
            setEmbeddingModel: (v) => { native_1.native.configSetEmbeddingModel(h, v); },
            setEmbeddingDimensions: (v) => { native_1.native.configSetEmbeddingDimensions(h, v); },
            setEmbeddingEndpoint: (v) => { native_1.native.configSetEmbeddingEndpoint(h, v); },
            setEmbeddingApiKey: (v) => { native_1.native.configSetEmbeddingApiKey(h, v); },
            setEmbeddingModelPath: (v) => { native_1.native.configSetEmbeddingModelPath(h, v); },
            setEmbeddingTokenizerPath: (v) => { native_1.native.configSetEmbeddingTokenizerPath(h, v); },
            // Vector DB
            setVectorDbProvider: (v) => { native_1.native.configSetVectorDbProvider(h, v); },
            setVectorDbUrl: (v) => { native_1.native.configSetVectorDbUrl(h, v); },
            setVectorDbKey: (v) => { native_1.native.configSetVectorDbKey(h, v); },
            setVectorDbHost: (v) => { native_1.native.configSetVectorDbHost(h, v); },
            setVectorDbPort: (v) => { native_1.native.configSetVectorDbPort(h, v); },
            setVectorDbName: (v) => { native_1.native.configSetVectorDbName(h, v); },
            // Graph DB
            setGraphDatabaseProvider: (v) => { native_1.native.configSetGraphDatabaseProvider(h, v); },
            setGraphModel: (v) => { native_1.native.configSetGraphModel(h, v); },
            setGraphFilePath: (v) => { native_1.native.configSetGraphFilePath(h, v); },
            // Chunking
            setChunkStrategy: (v) => { native_1.native.configSetChunkStrategy(h, v); },
            setChunkEngine: (v) => { native_1.native.configSetChunkEngine(h, v); },
            setChunkSize: (v) => { native_1.native.configSetChunkSize(h, v); },
            setChunkOverlap: (v) => { native_1.native.configSetChunkOverlap(h, v); },
            // Paths
            setSystemRootDirectory: (v) => { native_1.native.configSetSystemRootDirectory(h, v); },
            setDataRootDirectory: (v) => { native_1.native.configSetDataRootDirectory(h, v); },
            setCacheRootDirectory: (v) => { native_1.native.configSetCacheRootDirectory(h, v); },
            setLogsRootDirectory: (v) => { native_1.native.configSetLogsRootDirectory(h, v); },
            // Ontology
            setOntologyFilePath: (v) => { native_1.native.configSetOntologyFilePath(h, v); },
            setOntologyResolver: (v) => { native_1.native.configSetOntologyResolver(h, v); },
            setOntologyMatchingStrategy: (v) => { native_1.native.configSetOntologyMatchingStrategy(h, v); },
            // Other
            setMonitoringTool: (v) => { native_1.native.configSetMonitoringTool(h, v); },
            setClassificationModel: (v) => { native_1.native.configSetClassificationModel(h, v); },
            setSummarizationModel: (v) => { native_1.native.configSetSummarizationModel(h, v); },
            // Generic + bulk (fallible — wrap errors)
            set: (key, value) => {
                try {
                    native_1.native.configSet(h, key, value);
                }
                catch (e) {
                    throw (0, errors_1.wrapNativeError)(e);
                }
            },
            setLlmConfig: (values) => {
                try {
                    native_1.native.configSetLlmConfig(h, values);
                }
                catch (e) {
                    throw (0, errors_1.wrapNativeError)(e);
                }
            },
            setEmbeddingConfig: (values) => {
                try {
                    native_1.native.configSetEmbeddingConfig(h, values);
                }
                catch (e) {
                    throw (0, errors_1.wrapNativeError)(e);
                }
            },
            setVectorDbConfig: (values) => {
                try {
                    native_1.native.configSetVectorDbConfig(h, values);
                }
                catch (e) {
                    throw (0, errors_1.wrapNativeError)(e);
                }
            },
            setGraphDbConfig: (values) => {
                try {
                    native_1.native.configSetGraphDbConfig(h, values);
                }
                catch (e) {
                    throw (0, errors_1.wrapNativeError)(e);
                }
            },
            // Read-back
            get: () => configToCamel(native_1.native.getConfig(h)),
        };
        // ── datasets sub-object ──────────────────────────────────────────────────
        this.datasets = {
            list: () => native_1.native.cogneeListDatasets(h),
            listData: (datasetId) => native_1.native.cogneeListData(h, datasetId),
            has: (datasetId) => native_1.native.cogneeHasData(h, datasetId),
            status: (datasetIds) => native_1.native.cogneeDatasetStatus(h, datasetIds),
            empty: (datasetId) => native_1.native.cogneeEmptyDataset(h, datasetId),
            deleteData: (datasetId, dataId, opts) => native_1.native.cogneeDeleteData(h, datasetId, dataId, opts),
            deleteAll: () => native_1.native.cogneeDeleteAllDatasets(h),
        };
        // ── sessions sub-object ──────────────────────────────────────────────────
        this.sessions = {
            get: (sessionId, opts) => native_1.native.cogneeGetSession(h, sessionId, opts),
            addFeedback: (sessionId, qaId, text, score, opts) => native_1.native.cogneeAddFeedback(h, sessionId, qaId, text, score, opts),
            deleteFeedback: (sessionId, qaId, opts) => native_1.native.cogneeDeleteFeedback(h, sessionId, qaId, opts),
            getGraphContext: (sessionId, opts) => native_1.native.cogneeGetGraphContext(h, sessionId, opts),
            setGraphContext: (sessionId, context, opts) => native_1.native.cogneeSetGraphContext(h, sessionId, context, opts),
        };
        // ── notebooks sub-object ─────────────────────────────────────────────────
        this.notebooks = {
            list: () => native_1.native.cogneeListNotebooks(h),
            create: (name, cells, del) => native_1.native.cogneeCreateNotebook(h, name, cells, del),
            update: (id, patch) => native_1.native.cogneeUpdateNotebook(h, id, patch),
            delete: (id) => native_1.native.cogneeDeleteNotebook(h, id),
        };
        // ── users sub-object ─────────────────────────────────────────────────────
        this.users = {
            getOrCreateDefault: () => native_1.native.cogneeGetOrCreateDefaultUser(h),
            resetPipelineRunStatus: (datasetId, pipelineName) => native_1.native.cogneeResetPipelineRunStatus(h, datasetId, pipelineName),
            resetDatasetPipelineRunStatus: (datasetId) => native_1.native.cogneeResetDatasetPipelineRunStatus(h, datasetId),
        };
    }
    // ── Lifecycle ──────────────────────────────────────────────────────────────
    /**
     * Warm up the SDK: build embedding/LLM engines and resolve the default user.
     * Call this once after construction before running the pipeline.
     */
    async warm() {
        try {
            await native_1.native.cogneeWarm(this._handle);
        }
        catch (e) {
            throw (0, errors_1.wrapNativeError)(e);
        }
    }
    /** Return the UUID of the owner this handle is scoped to. */
    async ownerId() {
        try {
            return await native_1.native.cogneeOwnerId(this._handle);
        }
        catch (e) {
            throw (0, errors_1.wrapNativeError)(e);
        }
    }
    // ── Pipeline ops ──────────────────────────────────────────────────────────
    /**
     * Ingest data into a named dataset.
     *
     * `dataInput` is a single item or an array:
     * - `{ type: "text", text: "…" }`
     * - `{ type: "file", path: "/abs/path" }`
     * - `{ type: "url", url: "https://…" }`
     * - `{ type: "binary", bytes: Buffer, name: "doc.pdf" }`
     */
    async add(dataInput, datasetName, opts) {
        try {
            return await native_1.native.cogneeAdd(this._handle, dataInput, datasetName, opts);
        }
        catch (e) {
            throw (0, errors_1.wrapNativeError)(e);
        }
    }
    /**
     * Run the knowledge-graph extraction pipeline on a previously-added dataset.
     */
    async cognify(dataset, opts) {
        try {
            return await native_1.native.cogneeCognify(this._handle, dataset, opts);
        }
        catch (e) {
            throw (0, errors_1.wrapNativeError)(e);
        }
    }
    /**
     * Ingest data and immediately run knowledge-graph extraction in a single call.
     */
    async addAndCognify(dataInput, datasetName, opts) {
        try {
            return await native_1.native.cogneeAddAndCognify(this._handle, dataInput, datasetName, opts);
        }
        catch (e) {
            throw (0, errors_1.wrapNativeError)(e);
        }
    }
    // ── Retrieval ops ─────────────────────────────────────────────────────────
    /**
     * Search the knowledge graph.
     *
     * `opts.searchType` defaults to `"GRAPH_COMPLETION"`. All 15
     * SCREAMING_SNAKE_CASE variants are supported.
     */
    async search(query, opts) {
        try {
            return await native_1.native.cogneeSearch(this._handle, query, opts);
        }
        catch (e) {
            throw (0, errors_1.wrapNativeError)(e);
        }
    }
    /**
     * Recall information using session-first routing.
     *
     * Checks session QA history first (if `opts.sessionId` is set), then falls
     * back to graph search. Use `opts.scope` to control which sources contribute.
     */
    async recall(query, opts) {
        try {
            return await native_1.native.cogneeRecall(this._handle, query, opts);
        }
        catch (e) {
            throw (0, errors_1.wrapNativeError)(e);
        }
    }
    // ── Memory ops ────────────────────────────────────────────────────────────
    /**
     * Composite add+cognify with optional graph-improvement pass.
     *
     * `opts.selfImprovement` triggers a memify pass after cognify.
     * `opts.sessionId` switches to session-memory mode (no graph writes).
     */
    async remember(dataInput, datasetName, opts) {
        try {
            return await native_1.native.cogneeRemember(this._handle, dataInput, datasetName, opts);
        }
        catch (e) {
            throw (0, errors_1.wrapNativeError)(e);
        }
    }
    /**
     * Store a typed memory entry in a session.
     *
     * `entry` is a discriminated union: `"qa"`, `"trace"`, or `"feedback"`.
     */
    async rememberEntry(entry, datasetName, sessionId, opts) {
        try {
            return await native_1.native.cogneeRememberEntry(this._handle, entry, datasetName, sessionId, opts);
        }
        catch (e) {
            throw (0, errors_1.wrapNativeError)(e);
        }
    }
    /**
     * Index triplet embeddings from the existing knowledge graph.
     *
     * Idempotent (re-runnable). Enables `SearchType::TripletCompletion`.
     */
    async memify(opts) {
        try {
            return await native_1.native.cogneeMemify(this._handle, opts);
        }
        catch (e) {
            throw (0, errors_1.wrapNativeError)(e);
        }
    }
    /**
     * Run the four-stage session-graph bridge pipeline:
     * session collection → memify → feedback integration → edge synchronisation.
     */
    async improve(opts) {
        try {
            return await native_1.native.cogneeImprove(this._handle, opts);
        }
        catch (e) {
            throw (0, errors_1.wrapNativeError)(e);
        }
    }
    // ── Data ops ──────────────────────────────────────────────────────────────
    /**
     * Delete data scoped by the `target` discriminated union:
     * - `{ kind: "item", dataId, dataset }` — remove a single data item
     * - `{ kind: "dataset", dataset }` — remove an entire dataset
     * - `{ kind: "all" }` — remove everything
     */
    async forget(target, opts) {
        try {
            return await native_1.native.cogneeForget(this._handle, target, opts);
        }
        catch (e) {
            throw (0, errors_1.wrapNativeError)(e);
        }
    }
    /**
     * Replace an existing data item: delete it, re-add the new content, and
     * re-run knowledge-graph extraction.
     */
    async update(dataId, newData, datasetName, opts) {
        try {
            return await native_1.native.cogneeUpdate(this._handle, dataId, newData, datasetName, opts);
        }
        catch (e) {
            throw (0, errors_1.wrapNativeError)(e);
        }
    }
    /**
     * Remove all files from storage (non-cascading; metadata DB is not touched).
     */
    async pruneData() {
        try {
            await native_1.native.cogneePruneData(this._handle);
        }
        catch (e) {
            throw (0, errors_1.wrapNativeError)(e);
        }
    }
    /**
     * Selectively wipe graph, vector, metadata, and/or cache backends.
     *
     * All flags default to `true` when not specified.
     */
    async pruneSystem(opts) {
        try {
            return await native_1.native.cogneePruneSystem(this._handle, opts);
        }
        catch (e) {
            throw (0, errors_1.wrapNativeError)(e);
        }
    }
    // ── Visualization ─────────────────────────────────────────────────────────
    /**
     * Render the current knowledge graph as a self-contained d3.js HTML string.
     *
     * Throws `FeatureNotBuiltError` when the `visualization` feature was not
     * compiled into this build.
     */
    async visualize(opts) {
        try {
            return await native_1.native.cogneeVisualize(this._handle, opts);
        }
        catch (e) {
            throw (0, errors_1.wrapNativeError)(e);
        }
    }
    /**
     * Write the knowledge-graph visualization to an HTML file and return its
     * absolute path.
     *
     * `opts.destinationPath` overrides the default `~/graph_visualization.html`.
     *
     * Throws `FeatureNotBuiltError` when the `visualization` feature was not
     * compiled into this build.
     */
    async visualizeToFile(opts) {
        try {
            return await native_1.native.cogneeVisualizeToFile(this._handle, opts);
        }
        catch (e) {
            throw (0, errors_1.wrapNativeError)(e);
        }
    }
}
exports.Cognee = Cognee;
