// @cognee/cognee-ts/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.pipeline = exports.Cognee = exports.wrapNativeError = exports.ConfigTypeMismatchError = exports.UnknownConfigKeyError = exports.FeatureNotBuiltError = exports.UnsupportedError = exports.ValidationError = exports.RuntimeError = exports.UserBootstrapError = exports.ServiceBuildError = exports.ComponentError = exports.CogneeError = exports.RunHandle = exports.createNoopWatcher = exports.createWatcher = exports.Watcher = exports.ProgressToken = exports.createCancellationPair = exports.CancellationToken = exports.CancellationHandle = exports.Pipeline = exports.TaskContext = exports.createBatchTask = exports.createIterTask = exports.createTask = exports.TaskInfo = void 0;
exports.init = init;
exports.initWithThreads = initWithThreads;
exports.shutdown = shutdown;
exports.setupLogging = setupLogging;
exports.setupTelemetry = setupTelemetry;
exports.setupTelemetryAnalytics = setupTelemetryAnalytics;
/**
 * cognee — Node.js bindings for the cognee AI-memory SDK.
 *
 * Quick start:
 * ```ts
 * import { Cognee } from 'cognee';
 *
 * const c = new Cognee({ llmModel: "gpt-4o-mini" });
 * await c.warm();
 * await c.add({ type: "text", text: "…" }, "main");
 * await c.cognify("main");
 * const results = await c.search("my query");
 * ```
 *
 * Migration: Before: import { Pipeline } from '@cognee/pipeline'
 *            After:  import { pipeline } from 'cognee'
 *                    const { Pipeline } = pipeline
 *
 * Environment variables consumed by the Rust core on import:
 *
 *   COGNEE_BINDING_SUPPRESS_LOGS=1  — suppress the default
 *     tracing-subscriber stderr install (gap 07 decision 1). Set
 *     before `require`ing this module if your host owns its logger.
 *
 * After import, call `setupLogging()` to add file logging,
 * `setupTelemetry()` to add OTLP export, and
 * `setupTelemetryAnalytics()` to enable product-analytics emission.
 *
 * This module also attempts to load `.env` automatically (first from
 * `process.cwd()`, then from its parent directory). Set
 * `COGNEE_DISABLE_DOTENV=1` to disable auto-loading.
 */
require("./env");
const native_1 = require("./native");
// Runtime
function init() {
    native_1.native.init();
}
function initWithThreads(n) {
    native_1.native.initWithThreads(n);
}
function shutdown() {
    native_1.native.shutdown();
}
/**
 * Initialize cognee's logging subsystem from environment variables.
 *
 * All configuration is via env vars (`COGNEE_LOG_*`, `LOG_FILE_NAME`,
 * `LOG_LEVEL`, `RUST_LOG`); set them before calling. Calling this
 * function more than once is a no-op (idempotent).
 */
function setupLogging() {
    native_1.native.setupLogging();
}
/**
 * Initialize OpenTelemetry / OTLP export from environment variables.
 *
 * Reads `COGNEE_TRACING_ENABLED`, `OTEL_EXPORTER_OTLP_ENDPOINT`,
 * `OTEL_EXPORTER_OTLP_HEADERS`, `OTEL_SERVICE_NAME` and related
 * `OTEL_*` env vars. When neither `COGNEE_TRACING_ENABLED=true` nor a
 * non-empty `OTEL_EXPORTER_OTLP_ENDPOINT` is set, the call returns
 * successfully without installing anything (no-config = no-op).
 *
 * When `OTEL_SERVICE_NAME` is unset, defaults to `cognee.node-binding`
 * (gap 07 decision 8). The user's explicit value always wins.
 *
 * Calling more than once is a no-op (idempotent).
 */
function setupTelemetry() {
    native_1.native.setupTelemetry();
}
/**
 * Arm cognee product-analytics emission for this Node.js process.
 *
 * Default policy (gap 07 decision 11): ON unless `TELEMETRY_DISABLED`
 * is set, `ENV` is `"test"`/`"dev"`, or `COGNEE_HOST_SDK` is set —
 * Neon is the canonical sender in the JS ecosystem.
 *
 * Returns `true` if analytics were armed by this call (or a prior
 * call), `false` if the policy suppressed emission. Idempotent.
 */
function setupTelemetryAnalytics() {
    return native_1.native.setupTelemetryAnalytics();
}
var task_1 = require("./task");
Object.defineProperty(exports, "TaskInfo", { enumerable: true, get: function () { return task_1.TaskInfo; } });
Object.defineProperty(exports, "createTask", { enumerable: true, get: function () { return task_1.createTask; } });
Object.defineProperty(exports, "createIterTask", { enumerable: true, get: function () { return task_1.createIterTask; } });
Object.defineProperty(exports, "createBatchTask", { enumerable: true, get: function () { return task_1.createBatchTask; } });
var task_context_1 = require("./task-context");
Object.defineProperty(exports, "TaskContext", { enumerable: true, get: function () { return task_context_1.TaskContext; } });
var pipeline_1 = require("./pipeline");
Object.defineProperty(exports, "Pipeline", { enumerable: true, get: function () { return pipeline_1.Pipeline; } });
var cancellation_1 = require("./cancellation");
Object.defineProperty(exports, "CancellationHandle", { enumerable: true, get: function () { return cancellation_1.CancellationHandle; } });
Object.defineProperty(exports, "CancellationToken", { enumerable: true, get: function () { return cancellation_1.CancellationToken; } });
Object.defineProperty(exports, "createCancellationPair", { enumerable: true, get: function () { return cancellation_1.createCancellationPair; } });
var progress_1 = require("./progress");
Object.defineProperty(exports, "ProgressToken", { enumerable: true, get: function () { return progress_1.ProgressToken; } });
var watcher_1 = require("./watcher");
Object.defineProperty(exports, "Watcher", { enumerable: true, get: function () { return watcher_1.Watcher; } });
Object.defineProperty(exports, "createWatcher", { enumerable: true, get: function () { return watcher_1.createWatcher; } });
Object.defineProperty(exports, "createNoopWatcher", { enumerable: true, get: function () { return watcher_1.createNoopWatcher; } });
var run_handle_1 = require("./run-handle");
Object.defineProperty(exports, "RunHandle", { enumerable: true, get: function () { return run_handle_1.RunHandle; } });
var errors_1 = require("./errors");
Object.defineProperty(exports, "CogneeError", { enumerable: true, get: function () { return errors_1.CogneeError; } });
Object.defineProperty(exports, "ComponentError", { enumerable: true, get: function () { return errors_1.ComponentError; } });
Object.defineProperty(exports, "ServiceBuildError", { enumerable: true, get: function () { return errors_1.ServiceBuildError; } });
Object.defineProperty(exports, "UserBootstrapError", { enumerable: true, get: function () { return errors_1.UserBootstrapError; } });
Object.defineProperty(exports, "RuntimeError", { enumerable: true, get: function () { return errors_1.RuntimeError; } });
Object.defineProperty(exports, "ValidationError", { enumerable: true, get: function () { return errors_1.ValidationError; } });
Object.defineProperty(exports, "UnsupportedError", { enumerable: true, get: function () { return errors_1.UnsupportedError; } });
Object.defineProperty(exports, "FeatureNotBuiltError", { enumerable: true, get: function () { return errors_1.FeatureNotBuiltError; } });
Object.defineProperty(exports, "UnknownConfigKeyError", { enumerable: true, get: function () { return errors_1.UnknownConfigKeyError; } });
Object.defineProperty(exports, "ConfigTypeMismatchError", { enumerable: true, get: function () { return errors_1.ConfigTypeMismatchError; } });
Object.defineProperty(exports, "wrapNativeError", { enumerable: true, get: function () { return errors_1.wrapNativeError; } });
// ─────────────────────────────────────────────────────────────────────────────
// SDK-level exports (Phase 7)
// ─────────────────────────────────────────────────────────────────────────────
/** The main Cognee SDK class and sub-object types. Cloud module-level
 * functions (`serve` / `disconnect`) live in the closed `cognee-ts-cloud`
 * package (T15e). */
var cognee_1 = require("./cognee");
Object.defineProperty(exports, "Cognee", { enumerable: true, get: function () { return cognee_1.Cognee; } });
/** All user-facing Cognee* TypeScript types. */
__exportStar(require("./types"), exports);
/**
 * Low-level pipeline engine API (legacy).
 *
 * ```ts
 * import { pipeline } from 'cognee';
 * const p = new pipeline.Pipeline();
 * ```
 */
exports.pipeline = __importStar(require("./pipeline"));
