// @cognee/cognee-ts/lib/pipeline.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RunHandle = exports.createNoopWatcher = exports.createWatcher = exports.Watcher = exports.ProgressToken = exports.createCancellationPair = exports.CancellationToken = exports.CancellationHandle = exports.TaskContext = exports.createBatchTask = exports.createIterTask = exports.createTask = exports.TaskInfo = exports.Pipeline = void 0;
const native_1 = require("./native");
const task_1 = require("./task");
const run_handle_1 = require("./run-handle");
/** Builder for a cognee-core Pipeline. */
class Pipeline {
    constructor(description = "") {
        this._box = native_1.native.pipelineNew(description);
    }
    setName(name) {
        native_1.native.pipelineSetName(this._box, name);
        return this;
    }
    addTask(task) {
        // `createTask`/`createIterTask`/`createBatchTask` already return a
        // `TaskInfo`. The README also documents wrapping that result again with
        // `new TaskInfo(task)`, which nests one TaskInfo inside another so that
        // `_box` points at a JS `TaskInfo` instead of the native `NeonTaskInfo`
        // JsBox. Unwrap any such nesting so both usages reach the native box and
        // do not trigger a Neon downcast error.
        let box = task._box;
        while (box instanceof task_1.TaskInfo) {
            box = box._box;
        }
        native_1.native.pipelineAddTask(this._box, box);
        return this;
    }
    setBatchSize(size) {
        native_1.native.pipelineSetBatchSize(this._box, size);
        return this;
    }
    setConcurrency(n) {
        native_1.native.pipelineSetConcurrency(this._box, n);
        return this;
    }
    setRetry(policy) {
        native_1.native.pipelineSetRetry(this._box, policy);
        return this;
    }
    /**
     * Execute the pipeline synchronously (blocking worker thread).
     * Does NOT require `init()` — creates its own single-threaded tokio runtime.
     */
    async execute(inputs, ctx) {
        return native_1.native.pipelineExecute(this._box, inputs, ctx._box);
    }
    /**
     * Execute the pipeline asynchronously on the global tokio runtime.
     * Requires `init()` to have been called.
     */
    async executeAsync(inputs, ctx) {
        return native_1.native.pipelineExecuteAsync(this._box, inputs, ctx._box);
    }
    /**
     * Execute the pipeline in the background. Returns a handle immediately.
     * Requires `init()` to have been called.
     */
    executeInBackground(inputs, ctx) {
        const handle = native_1.native.pipelineExecuteBackground(this._box, inputs, ctx._box);
        return new run_handle_1.RunHandle(handle);
    }
    /**
     * Execute with a watcher for lifecycle event callbacks.
     * Requires `init()` to have been called.
     */
    async executeWithWatcher(inputs, ctx, watcher) {
        return native_1.native.pipelineExecuteWithWatcher(this._box, inputs, ctx._box, watcher._box);
    }
}
exports.Pipeline = Pipeline;
var task_2 = require("./task");
Object.defineProperty(exports, "TaskInfo", { enumerable: true, get: function () { return task_2.TaskInfo; } });
Object.defineProperty(exports, "createTask", { enumerable: true, get: function () { return task_2.createTask; } });
Object.defineProperty(exports, "createIterTask", { enumerable: true, get: function () { return task_2.createIterTask; } });
Object.defineProperty(exports, "createBatchTask", { enumerable: true, get: function () { return task_2.createBatchTask; } });
var task_context_1 = require("./task-context");
Object.defineProperty(exports, "TaskContext", { enumerable: true, get: function () { return task_context_1.TaskContext; } });
var cancellation_1 = require("./cancellation");
Object.defineProperty(exports, "CancellationHandle", { enumerable: true, get: function () { return cancellation_1.CancellationHandle; } });
Object.defineProperty(exports, "CancellationToken", { enumerable: true, get: function () { return cancellation_1.CancellationToken; } });
Object.defineProperty(exports, "createCancellationPair", { enumerable: true, get: function () { return cancellation_1.createCancellationPair; } });
var progress_1 = require("./progress");
Object.defineProperty(exports, "ProgressToken", { enumerable: true, get: function () { return progress_1.ProgressToken; } });
var watcher_1 = require("./watcher");
Object.defineProperty(exports, "Watcher", { enumerable: true, get: function () { return watcher_1.Watcher; } });
Object.defineProperty(exports, "createWatcher", { enumerable: true, get: function () { return watcher_1.createWatcher; } });
Object.defineProperty(exports, "createNoopWatcher", { enumerable: true, get: function () { return watcher_1.createNoopWatcher; } });
var run_handle_2 = require("./run-handle");
Object.defineProperty(exports, "RunHandle", { enumerable: true, get: function () { return run_handle_2.RunHandle; } });
