// @cognee/cognee-ts/lib/progress.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProgressToken = void 0;
const native_1 = require("./native");
class ProgressToken {
    /** @internal */
    constructor(box_) {
        this._box = box_;
    }
    /** Create a root progress token at 0%. */
    static create() {
        return new ProgressToken(native_1.native.progressNew());
    }
    /** Set this token's progress fraction (clamped to [0, 1]). */
    set(fraction) {
        native_1.native.progressSet(this._box, fraction);
    }
    /** This token's progress fraction in [0, 1]. */
    get fraction() {
        return native_1.native.progressFraction(this._box);
    }
    /** This token's width as a fraction of the root [0, 1] range. */
    get width() {
        return native_1.native.progressWidth(this._box);
    }
    /** Whether this token's progress >= 1.0. */
    get isComplete() {
        return native_1.native.progressIsComplete(this._box);
    }
    /** Overall progress across the entire tree. */
    get rootFraction() {
        return native_1.native.progressRootFraction(this._box);
    }
    /** Split into subtokens by relative weights. */
    split(weights) {
        const boxes = native_1.native.progressSplit(this._box, weights);
        return boxes.map((b) => new ProgressToken(b));
    }
    /** Create one child subtoken covering `fracWidth` of this token's range. */
    subtoken(fracWidth) {
        return new ProgressToken(native_1.native.progressSubtoken(this._box, fracWidth));
    }
    clone() {
        return new ProgressToken(native_1.native.progressClone(this._box));
    }
}
exports.ProgressToken = ProgressToken;
